anonymous_helpline_chatbot/core/invitations.py

```diff
@@ -1,89 +1,131 @@
-        :param tg_client_id: Telegram identifier of the user to invite operators to chat with
-        :return: Error code, either of `0`, `1`, `2`, `3`, where `0` indicates that the invitations have been sent
-            successfully, `1` tells that the user had requested invitations before, `2` indicates that there are no free
-            operators, `3` means that the client is in a conversation already (either as a client or as an operator)
+        :param client_chat_id: Messenger identifier of the user to invite operators to have conversation with
         """
-        local_client_id = self.users_controller.get_local_id(tg_client_id)
-
-        with self.conversations_controller.conversations_starter_finisher_lock:
-            free_operators = set(self.users_controller.get_free_operators()) - {tg_client_id}
-            if not free_operators:
-                return 2
-
-            if tg_client_id in self._operators_invitations_messages.keys():
-                return 1
-
-            # In a conversation already
-            if self.conversations_controller.get_conversing(tg_client_id) != ((None, None), (None, None)):
-                return 3
-
-            msg_ids = []
-            for tg_operator_id in free_operators:
-                try:
-                    msg_ids.append((
-                        tg_operator_id,
-                        self.send_invitation_callback(tg_operator_id, tg_client_id,
-                                                      f"Пользователь №{local_client_id} хочет побеседовать. Нажмите "
-                                                      "кнопку ниже, чтобы стать его оператором")
-                    ))
-                except telebot.apihelper.ApiException:
-                    print("Telegram API Exception while sending out operators invitations:", file=stderr)
-                    print(format_exc(), file=stderr)
-
-            self._operators_invitations_messages[tg_client_id] = msg_ids
-
-        return 0
-
-    def clear_invitation_messages(self, tg_client_id: int) -> bool:
+        with self._conn_pool.PrettyCursor() as cursor:
+            # Prevent any invitations from being sent or deleted by parallel transactions until this one completes,
+            # because, for example, a parallel transaction might try to delete an invitation which we are going to send,
+            # but have not sent yet
+            cursor.execute("LOCK TABLE sent_invitations IN SHARE MODE")
+
+            """
+            Explanation of the query below:
+
+            This query selects chat ids of all the operators which should receive an invitation to the given client
+            (i.e. they are currently not in conversations, and they haven't yet been sent an invitation to the client).
+            We select them with the two joins in the following way (of course, PostgreSQL doesn't do exactly what I say
+            here, it performs optimizations. But the result is exactly as if it was doing the following):
```

Comment:
```suggestion
            here, it performs optimizations. But the result is exactly as if it would be if pgsql was
            doing the following):
```